File: budget-pipeline/scrubber.py
"""
PII scrubber for merchant names.
Removes personal names, account numbers, reference codes, and other
identifiable information from merchant strings before storage.
"""

import re


def scrub(merchant: str, description: str, ach_names: list[str]) -> str:
    """
    Scrub PII and noise from a merchant name.

    Rules applied in order:
    1. Zelle payments: strip names → "Zelle Payment"
    2. ACH known names: replace whole merchant → "ACH Transfer"
    3. PPD ID / WEB ID / ACH ID suffixes: strip them
    4. Check numbers: → "Personal Check"
    5. Trailing reference codes (8+ alphanumeric): strip them
    6. Strip partial account numbers like -71005
    7. Collapse whitespace, title-case, truncate to 28 chars
    """
    s = merchant.strip()

    # Rule 1: Zelle payments — strip personal names
    # Match "Zelle Payment To John Smith" or "Zelle From Jane Doe" etc.
    if re.search(r'(?i)zelle', s):
        if re.search(r'(?i)zelle\s+(payment\s+)?(to|from)\s+[A-Za-z][A-Za-z\s]+$', s):
            return "Zelle Payment"
        # Also check description for zelle from/to patterns
        if re.search(r'(?i)zelle\s+(payment\s+)?(to|from)\s+[A-Za-z][A-Za-z\s]+$', description):
            return "Zelle Payment"
        # Zelle present but no name pattern — still clean to "Zelle Payment"
        # only if it's clearly a payment
        if re.search(r'(?i)zelle\s+(payment|transfer)', s):
            return "Zelle Payment"

    # Rule 2: ACH known names — replace whole merchant with "ACH Transfer"
    s_upper = s.upper()
    for name in ach_names:
        if name.strip().upper() in s_upper:
            return "ACH Transfer"

    # Rule 3: Strip PPD ID / WEB ID / ACH ID / TEL ID suffixes.
    # Digits may already be stripped by clean_merchant, so make the digit group optional.
    s = re.sub(r'\s+PPD\s+ID[:\s]*\d*', '', s, flags=re.IGNORECASE).strip()
    s = re.sub(r'\s+WEB\s+ID[:\s]*\d*', '', s, flags=re.IGNORECASE).strip()
    s = re.sub(r'\s+ACH\s+ID[:\s]*\d*', '', s, flags=re.IGNORECASE).strip()
    s = re.sub(r'\s+TEL\s+ID[:\s]*\d*', '', s, flags=re.IGNORECASE).strip()
    # Strip any trailing isolated colon left after digit removal
    s = re.sub(r'[\s:]+$', '', s).strip()

    # Rule 4: Check numbers → "Personal Check"
    if re.search(r'(?i)\bcheck\s+\d+', s):
        return "Personal Check"

    # Rule 5: Trailing reference codes (8+ alphanumeric chars)
    s = re.sub(r'\s+[A-Z0-9]{8,}$', '', s).strip()

    # Rule 6: Strip partial account numbers (hyphen + digits)
    s = re.sub(r'-\d+', '', s).strip()

    # Rule 7: Collapse whitespace, title-case, truncate to 28 chars
    s = re.sub(r'\s+', ' ', s).strip()
    s = s.title()
    s = s[:28]

    return s

File: web/lib/import/scrubber.ts
// Faithful TS port of budget-pipeline/scrubber.py :: scrub.
// Removes PII/noise from merchant names before storage. Rule order matters.

import { pythonTitle } from './clean-merchant'

const ZELLE_NAME_RE = /zelle\s+(payment\s+)?(to|from)\s+[A-Za-z][A-Za-z\s]+$/i

/**
 * Scrub PII and noise from a merchant name.
 * @param merchant   cleaned or raw merchant string
 * @param description raw bank description (used as a fallback signal for Zelle)
 * @param achNames    personal names to mask as "ACH Transfer"
 */
export function scrub(merchant: string, description: string, achNames: string[]): string {
  let s = merchant.trim()

  // Rule 1: Zelle payments — strip personal names
  if (/zelle/i.test(s)) {
    if (ZELLE_NAME_RE.test(s)) return 'Zelle Payment'
    if (ZELLE_NAME_RE.test(description)) return 'Zelle Payment'
    if (/zelle\s+(payment|transfer)/i.test(s)) return 'Zelle Payment'
  }

  // Rule 2: ACH known names — replace whole merchant
  const sUpper = s.toUpperCase()
  for (const name of achNames) {
    if (name.trim() && sUpper.includes(name.trim().toUpperCase())) {
      return 'ACH Transfer'
    }
  }

  // Rule 3: Strip PPD / WEB / ACH / TEL ID suffixes (digits may already be gone)
  s = s.replace(/\s+PPD\s+ID[:\s]*\d*/gi, '').trim()
  s = s.replace(/\s+WEB\s+ID[:\s]*\d*/gi, '').trim()
  s = s.replace(/\s+ACH\s+ID[:\s]*\d*/gi, '').trim()
  s = s.replace(/\s+TEL\s+ID[:\s]*\d*/gi, '').trim()
  // Strip any trailing isolated colon/space left after digit removal
  s = s.replace(/[\s:]+$/, '').trim()

  // Rule 4: Check numbers -> "Personal Check"
  if (/\bcheck\s+\d+/i.test(s)) return 'Personal Check'

  // Rule 5: Trailing reference codes (8+ uppercase alphanumerics)
  s = s.replace(/\s+[A-Z0-9]{8,}$/, '').trim()

  // Rule 6: Strip partial account numbers (hyphen + digits)
  s = s.replace(/-\d+/g, '').trim()

  // Rule 7: Collapse whitespace, title-case, truncate to 28 chars
  s = s.replace(/\s+/g, ' ').trim()
  s = pythonTitle(s)
  s = s.slice(0, 28)

  return s
}
